wgengine/netstack/netstack.go

```diff
@@ -328,19 +387,30 @@ func (ns *Impl) acceptTCP(r *tcp.ForwarderRequest) {
 		r.Complete(true)
 		return
 	}
-	localAddr, err := ep.GetLocalAddress()
+	dialAddr, err := ep.GetLocalAddress()
 	if err != nil {
 		r.Complete(true)
 		return
 	}
+	dialNetAddr, _ := netaddr.FromStdIP(net.IP(dialAddr.Addr))
+	isTailscaleIP := tsaddr.IsTailscaleIP(dialNetAddr)
+	if isTailscaleIP {
+		dialAddr.Addr = tcpip.Address(net.ParseIP("127.0.0.1")).To4()
+	}
 	r.Complete(false)
 	c := gonet.NewTCPConn(&wq, ep)
-	go ns.forwardTCP(c, &wq, localAddr.Port)
+	ns.forwardTCP(c, &wq, dialAddr)
+	if !isTailscaleIP {
+		// if this is a subnet IP, we added this in before the TCP handshake
+		// so netstack is happy TCP-handshaking as a subnet IP
+		ns.removeSubnetAddress(dialAddr.Addr)
```

Comment:
there are two other exit paths up there where you don't call this. I'd put this in a defer at the very top so it can't get lost.

The defer func could close over the `isTailscaleIP` variable, which you'd also need to move (the declaration of) to the top. You could still initialize it where it is.

But is this reference counting correct? Aren't you always incrementing the count on line 135? But here you only sometimes decrement it?